src/LanguageServer/Impl/Indexing/IndexManager.cs

```diff
@@ -0,0 +1,140 @@
+﻿// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABILITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Python.Analysis.Core.Interpreter;
+using Microsoft.Python.Analysis.Documents;
+using Microsoft.Python.Core;
+using Microsoft.Python.Core.Diagnostics;
+using Microsoft.Python.Core.Idle;
+using Microsoft.Python.Core.IO;
+using Microsoft.Python.Parsing;
+
+namespace Microsoft.Python.LanguageServer.Indexing {
+    internal class IndexManager : IIndexManager {
+        private const int DefaultReIndexDelay = 350;
+        private readonly ISymbolIndex _symbolIndex;
+        private readonly IFileSystem _fileSystem;
+        private readonly string _workspaceRootPath;
+        private readonly string[] _includeFiles;
+        private readonly string[] _excludeFiles;
+        private readonly IIdleTimeService _idleTimeService;
+        private readonly PythonLanguageVersion _version;
+        private readonly ConcurrentDictionary<IDocument, DateTime> _pendingDocs = new ConcurrentDictionary<IDocument, DateTime>(new UriDocumentComparer());
+
+        public IndexManager(ISymbolIndex symbolIndex, IFileSystem fileSystem, PythonLanguageVersion version, string rootPath, string[] includeFiles,
+            string[] excludeFiles, IIdleTimeService idleTimeService) {
+            Check.ArgumentNotNull(nameof(fileSystem), fileSystem);
+            Check.ArgumentNotNull(nameof(rootPath), rootPath);
+            Check.ArgumentNotNull(nameof(includeFiles), includeFiles);
+            Check.ArgumentNotNull(nameof(excludeFiles), excludeFiles);
+            Check.ArgumentNotNull(nameof(idleTimeService), idleTimeService);
+
+            _symbolIndex = symbolIndex;
+            _fileSystem = fileSystem;
+            _workspaceRootPath = rootPath;
+            _includeFiles = includeFiles;
+            _excludeFiles = excludeFiles;
+            _idleTimeService = idleTimeService;
+            _idleTimeService.Idle += OnIdle;
```

Comment:
Need to remove this in `Dispose()`.

Reply:
Have a look how `DisposableBag` is used

Reply:
If you decide to set up a bag in the constructor, you can also remove `_idleTimeService` from the class, as it would have no references other than `Idle +=` and `Idle -=`.